proposals/open-telemetry/microprofile-opentelemetry-tracing.asciidoc

```diff
@@ -0,0 +1,190 @@
+distributed tracing by using the https://opentelemetry.io[OpenTelemetry project] (a merger of the OpenTracing and
+OpenCensus projects).
+
+This document and implementations *MUST* comply with the following OpenTelemetry specifications:
+
+- https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/overview.md[OpenTelemetry Overview]
+(except for Metrics)
+- https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md[Tracing API]
+- https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/baggage/api.md[Baggage API]
+- https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/context[Context API]
+- https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md[Resource SDK]
+
+=== Automatic Instrumentation
+
+JAX-RS applications, JAX-RS Clients, and MicroProfile REST Clients are automatically enlisted to participate in
+distributed tracing without code modification.
+
+=== Manual Instrumentation
+
+Explicit manual instrumentation can be added into a MicroProfile application in the following ways:
+
+==== @WithSpan
+
+Annotating a method in any Jakarta CDI aware beans with the `io.opentelemetry.extension.annotations.WithSpan`
+annotation. This will create a new Span and establish any required relationships with the current Trace context.
+
+Method parameters can be annotated with the `io.opentelemetry.extension.annotations.SpanAttribute` annotation to
+indicate which method parameters should be part of the Trace.
+
+Example:
+[source,java]
+----
+@ApplicationScoped
+class SpanBean {
+    @WithSpan
+    void span() {
+
+    }
+
+    @WithSpan("name")
+    void spanName() {
+
+    }
+
+    @WithSpan(kind = SERVER)
+    void spanKind() {
+
+    }
+
+    @WithSpan
+    void spanArgs(@SpanAttribute(value = "arg") String arg) {
+
+    }
+}
+----
+
+==== Obtain a SpanBuilder
+
+By obtaining a `SpanBuilder` from the current `Tracer` and calling
+`io.opentelemetry.api.trace.Tracer.spanBuilder(String)`. In this case, it is the developer's responsibility to ensure
+that the `Span` is properly created, closed, and propagated.
+
+Example:
+[source,java]
+----
+@RequestScoped
+@Path("/")
+public class SpanResource {
+    @Inject
+    Tracer tracer;
+
+    @GET
+    @Path("/span/new")
+    public Response spanNew() {
+        Span span = tracer.spanBuilder("span.new")
+                .setSpanKind(INTERNAL)
+                .setParent(Context.current().with(this.span))
+                .setAttribute("my.attribute", "value")
+                .startSpan();
+
+        span.end();
+
+        return Response.ok().build();
+    }
+}
+----
+
+==== Obtain the current Span
+
+By obtaining the current `Span` to add attributes. The Span lifecycle is managed by the implementation.
+
+Example:
+[source,java]
+----
+@RequestScoped
+@Path("/")
+public class SpanResource {
+    @Inject
+    Span span;
```

Comment:
The current span can be obtained from `Span.current()`.  Do we still need inject it?

Reply:
A CDI Producer would just call `Span.current()`. The idea was just to align with the usual CDI programming model. We can remove that requirement and just use the static method calls to get the OTel objects.

Reply:
We discussed on today's call. We think calling `Span.current()` is simplier. Are you ok to update the example to reflect that?

Reply:
Is this just for the example, or do we want to drop the ability the `Span` with CDI?

Reply:
Added both programatic and CDI  example. Clarified that the OTel programmatic API can also be used.